src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs

```diff
@@ -800,7 +800,9 @@ public static unsafe T[] AllocateUninitializedArray<T>(int length, bool pinned =
             if (pinned)
                 flags |= GC_ALLOC_FLAGS.GC_ALLOC_PINNED_OBJECT_HEAP;
 
-            return Unsafe.As<T[]>(AllocateNewArray(RuntimeTypeHandle.ToIntPtr(typeof(T[]).TypeHandle), length, flags));
+            T[]? result = null;
+            AllocateNewArray(RuntimeTypeHandle.ToIntPtr(typeof(T[]).TypeHandle), length, flags, ObjectHandleOnStack.Create(ref result));
```

Comment:
I see ~10x perf regression in microbenchmark for StringBuilder usage pattern of this API:
```
using System;
using System.Runtime.CompilerServices;
using System.Diagnostics;

var sw = new Stopwatch();
for (;;)
{
    sw.Restart();
    for (int i = 0; i < 100000000; i++)
        Work<char>(1);
    Console.WriteLine(sw.ElapsedMilliseconds.ToString());
}

[MethodImpl(MethodImplOptions.NoInlining)]
static void Work<T>(int size)
{
   GC.KeepAlive(GC.AllocateArray<T>(size));
}
```
We should split it into fast/slow methods to avoid the perf regression. NativeAOT does something similar: https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/System.Private.CoreLib/src/System/GC.NativeAot.cs#L815-L816

Reply:
> I see ~10x perf regression in microbenchmark for StringBuilder usage pattern of this API:

Hmmm. That is basically the code I wrote. I must have done something wrong. Let me try this again.

Reply:
I copied a wrong copy of the microbenchmark. It should be testing `GC.KeepAlive(GC.AllocateUninitializedArray<T>(size));` (It is what StringBuilder uses.)

Reply:
EgorBot measured 1.43x regression for microbenchmark with fewer iterations.

Reply:
and [2X on arm64 ](https://github.com/EgorBot/runtime-utils/issues/16#issuecomment-2323514086)